projects/packages/stats/src/class-main.php

```diff
@@ -0,0 +1,217 @@
+use Automattic\Jetpack\Status\Visitor;
+use WP_User;
+
+/**
+ * Stats Main class.
+ *
+ * Entrypoint for Stats.
+ *
+ * @since $$next-version$$
+ */
+class Main {
+
+	/**
+	 * Singleton Main instance.
+	 *
+	 * @var Main
+	 **/
+	private static $instance = null;
+
+	/**
+	 * Initializer.
+	 *
+	 * @return object
+	 */
+	public static function init() {
+		if ( null === self::$instance ) {
+			self::$instance = new Main();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Class constructor.
+	 *
+	 * @return void
+	 */
+	private function __construct() {
+		defined( 'STATS_VERSION' ) || define( 'STATS_VERSION', '9' );
+		defined( 'STATS_DASHBOARD_SERVER' ) || define( 'STATS_DASHBOARD_SERVER', 'dashboard.wordpress.com' );
+		// Generate the tracking code after wp() has queried for posts.
+		add_action( 'template_redirect', array( __CLASS__, 'template_redirect' ), 1 );
+
+		add_action( 'wp_head', array( __CLASS__, 'hide_smile_css' ) );
+		add_action( 'embed_head', array( __CLASS__, 'hide_smile_css' ) );
+
+		// Map stats caps.
+		add_filter( 'map_meta_cap', array( __CLASS__, 'map_meta_caps' ), 10, 3 );
+
+		XMLRPC::init();
+	}
+
+	/**
+	 * Checks if filter is set and dnt is enabled.
+	 *
+	 * @return bool
+	 */
+	public static function jetpack_is_dnt_enabled() {
+		/**
+		 * Filter the option which decides honor DNT or not.
+		 *
+		 * @module stats
+		 * @since-jetpack 6.1.0
+		 *
+		 * @param bool false Honors DNT for clients who don't want to be tracked. Defaults to false. Set to true to enable.
+		 */
+		if ( false === apply_filters( 'jetpack_honor_dnt_header_for_stats', false ) ) {
+			return false;
+		}
+
+		foreach ( $_SERVER as $name => $value ) {
+			if ( 'http_dnt' === strtolower( $name ) && 1 === (int) $value ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Maps view_stats cap to read cap as needed.
+	 *
+	 * @access public
+	 * @param mixed $caps Caps.
+	 * @param mixed $cap Cap.
+	 * @param mixed $user_id User ID.
+	 * @return array Possibly mapped capabilities for meta capability.
+	 */
+	public static function map_meta_caps( $caps, $cap, $user_id ) {
+		// Map view_stats to exists.
+		if ( 'view_stats' === $cap ) {
+			$user        = new WP_User( $user_id );
+			$user_role   = array_shift( $user->roles );
+			$stats_roles = Options::get_option( 'roles' );
+
+			// Is the users role in the available stats roles?
+			if ( is_array( $stats_roles ) && in_array( $user_role, $stats_roles, true ) ) {
+				$caps = array( 'read' );
+			}
+		}
+
+		return $caps;
+	}
+
+	/**
+	 * Stats Template Redirect.
+	 *
+	 * @access public
+	 * @return void
+	 */
+	public static function template_redirect() {
+		global $current_user;
+
+		if (
+			is_feed()
+			|| is_robots()
+			|| is_embed()
+			|| is_trackback()
+			|| is_preview()
+			|| self::jetpack_is_dnt_enabled()
+		) {
+			return;
+		}
+
+		// @todo Is it ok to assume tracking will be enabled for any connected site?
+		if ( ! ( new Connection_Manager() )->is_connected() ) {
```

Comment:
I added this check now. My assumption was that the tracking pixel will be added no matter if the Stats module is enable therefore it only needs a connected site to work. We need to think if establishing a site connection is enough to start tracking users.
Up to now, when the module is disabled we stop adding the tracking pixel.. I'm not sure if we want to keep this functionality.
For example, Search might want to track pageviews no matter if the module is enabled or not (assuming both Search and Jetpack plugin coexist)